### PR TITLE
Fix handling of empty response bodies.

### LINKED ISSUE #66: an empty body will result in the response being ignore (creating no event)
you can use the following config to verify:

```
input {
  http_poller {
    urls => {
      test_404 => "http://www.mocky.io/v2/5873725d0f0000250452e130"
      test_ok => "http://www.mocky.io/v2/5873717f0f0000180452e12b"
      test_503 => "http://www.mocky.io/v2/587371e50f0000240452e12d"
      test_503_no_body => "http://www.mocky.io/v2/5873738c0f0000510452e134"
    }
    request_timeout => 60
    schedule => { cron => "* * * * * UTC"}
  }
}
output {
  stdout { codec => rubydebug }
}
```

the `test_503_no_body` will not generate an event. this is because: the response is ignore if it contains no body. This is because [this block of code](https://github.com/logstash-plugins/logstash-input-http_poller/blob/master/lib/logstash/inputs/http_poller.rb#L243-L246):

```ruby
  def handle_success(queue, name, request, response, execution_time)
    @codec.decode(response.body) do |decoded|
      event = @target ? LogStash::Event.new(@target => decoded.to_hash) : decoded
      handle_decoded_event(queue, name, request, response, event, execution_time)
    end
end
```

In response codes 200, 404, 503, etc, the `handle_success` is called and the body is decoded. The resulting event is created in the decode block of the codec, but since the body is empty, the codec doesn't yield any data, thus no event is created. I believe this to be a bug.

### PR DESCRIPTION
**Pull request by andrewvc**
These now emit an empty event with the poller metadata included

Fixes #66

### CODE REVIEW

File: lib/logstash/inputs/http_poller.rb [hunk 1]
```diff
@@ -240,12 +240,20 @@ def request_async(queue, name, request)
 
   private
   def handle_success(queue, name, request, response, execution_time)
-    @codec.decode(response.body) do |decoded|
-      event = @target ? LogStash::Event.new(@target => decoded.to_hash) : decoded
+    body = response.body
+    # If there is a usable response. HEAD requests are `nil` and empty get
+    # responses come up as "" which the codec cannot deal with
+    if body && body.size > 0
+      @codec.decode(response.body) do |decoded|
```
Comment (jsvd) [inline lib/logstash/inputs/http_poller.rb]: if we're extracting the body on line 243, I suggest using that here too

File: lib/logstash/inputs/http_poller.rb [hunk 2]
```diff
@@ -240,12 +240,20 @@ def request_async(queue, name, request)
 
   private
   def handle_success(queue, name, request, response, execution_time)
-    @codec.decode(response.body) do |decoded|
-      event = @target ? LogStash::Event.new(@target => decoded.to_hash) : decoded
+    body = response.body
+    # If there is a usable response. HEAD requests are `nil` and empty get
+    # responses come up as "" which the codec cannot deal with
```
Comment (jsvd) [inline lib/logstash/inputs/http_poller.rb]: suggestion:

s/which the codec will not yield anything

File: spec/inputs/http_poller_spec.rb [hunk 3]
```diff
@@ -395,9 +395,10 @@
         include_examples("unprocessable_requests")
       end
     end
-
+
```
Comment (jsvd) [inline spec/inputs/http_poller_spec.rb]: unnecessary whitespace

File: lib/logstash/inputs/http_poller.rb [hunk 4]
```diff
@@ -240,12 +240,20 @@ def request_async(queue, name, request)
...
+    if body && body.size > 0
+      @codec.decode(response.body) do |decoded|
+        event = @target ? LogStash::Event.new(@target => decoded.to_hash) : decoded
+        handle_decoded_event(queue, name, request, response, event, execution_time)
+      end
+    else
+      event = ::LogStash::Event.new
```
Comment (jsvd) [inline lib/logstash/inputs/http_poller.rb]: this control flow means that a 503 will be treated as a sucessful request? we call `handle_decoded_event` next which means `decorate` will be called, etc?
not sure if this is the ideal behaviour, thoughts?

Comment (jsvd) [inline lib/logstash/inputs/http_poller.rb]: I know the above comment seems out of scope for this PR, but a reminder that in the past we were dropping these events, and now we're going to start producing them, so we need to be careful on how they're crafted.

File: lib/logstash/inputs/http_poller.rb [hunk 5]
```diff
@@ -240,12 +240,20 @@ def request_async(queue, name, request)
...
       handle_decoded_event(queue, name, request, response, event, execution_time)
     end
   end
-
+
```
Comment (jsvd) [inline lib/logstash/inputs/http_poller.rb]: unnecessary whitespace

File: lib/logstash/inputs/http_poller.rb (hunk 4 above)
Comment (andrewvc) [inline lib/logstash/inputs/http_poller.rb]: I don't think the handling of 503s has changed at all. 503s can have bodies, so I don't think this impacts that use case.

Comment (andrewvc) [inline lib/logstash/inputs/http_poller.rb]: In fact most 503s generated by popular web servers *do* have bodies, I doubt this will affect anyone.

**Comment by jsvd** [comment]
LGTM, tested manually too, works perfectly 👍
